### PR TITLE
Add Support For Updating Newsletter Subscription In All Stores

### PR DESCRIPTION
**Pull request by supercid**

### CODE REVIEW

File: classes/services/NostoCustomerService.php [hunk 1]
```diff
@@ -39,38 +39,91 @@ class NostoCustomerService extends AbstractNostoService
...
+            ) {
+                try {
+                    $account = NostoHelperAccount::getAccount();
+                    if (!$account instanceof NostoSDKAccount || !$account->isConnectedToNosto()) {
+                        return;
+                    }
+                    $this->updateMarketingPermissionInCurrentContext($updatedAccounts, $customer, $account);
+                } catch (\Exception $e) {
+                    NostoHelperLogger::error($e);
+                    return;
```
Comment (mridang) [inline classes/services/NostoCustomerService.php]: Needed?

File: classes/services/NostoCustomerService.php [hunk 2]
```diff
@@ -39,38 +39,91 @@ class NostoCustomerService extends AbstractNostoService
...
+        NostoHelperContext::runInContextForEachLanguageEachShop(function () use ($customer, &$updatedAccounts) {
+            $shopGroup = Shop::getContextShopGroup();
+            if ($shopGroup !== null
+                && (bool)$shopGroup->active
+                && (bool)$shopGroup->share_customer
+            ) {
+                try {
+                    $account = NostoHelperAccount::getAccount();
+                    if (!$account instanceof NostoSDKAccount || !$account->isConnectedToNosto()) {
+                        return;
```
Comment (mridang) [inline classes/services/NostoCustomerService.php]: Maybe invert condition?

File: classes/services/NostoCustomerService.php (hunk 1 above)
Comment (supercid) [reply classes/services/NostoCustomerService.php]: Yes, `NostoHelperAccount::getAccount();` Throws an exception

File: classes/services/NostoCustomerService.php (hunk 2 above)
Comment (supercid) [reply classes/services/NostoCustomerService.php]: Fixed, thanks

File: classes/services/NostoCustomerService.php (hunk 1 above)
Comment (supercid) [reply classes/services/NostoCustomerService.php]: Oh, you mean the return statement. Not necessary, just removed

File: classes/services/NostoCustomerService.php [hunk 3]
```diff
@@ -39,38 +39,89 @@ class NostoCustomerService extends AbstractNostoService
...
 
-            if (!$account->getApiToken(NostoSDKToken::API_EMAIL)) {
+    /**
+     * Check if all store views that the customer belongs to
+     * were updated. Logs stores that failed to update.
+     *
+     * @param $updatedAccounts
+     * @return bool
+     */
+    private function isAllUpdated($updatedAccounts)
```
Comment (hannunosto) [inline classes/services/NostoCustomerService.php]: Typehint with `array` -> `isAllUpdated(array $updatedAccounts)`

File: classes/services/NostoCustomerService.php [hunk 4]
```diff
@@ -39,38 +39,89 @@ class NostoCustomerService extends AbstractNostoService
...
                     sprintf(
-                        "API_EMAIL api token is missing (%s). Please reconnect nosto account to create API_EMAIL token",
-                        $account->getName()
+                        'Failed to update marketing permission for the account (%s) ',
+                        $accountName
                     )
                 );
-
-                return false;
+                $success = false;
```
Comment (hannunosto) [inline classes/services/NostoCustomerService.php]: Not a biggie but you could `break` here

File: classes/services/NostoCustomerService.php [hunk 5]
```diff
@@ -39,38 +39,89 @@ class NostoCustomerService extends AbstractNostoService
...
+        }
+        return $this->isAllUpdated($updatedAccounts);
+    }
+
+    /**
+     * @param $updatedAccounts
+     * @param $customer
+     * @param NostoSDKAccount $account
+     */
+    private function updateMarketingPermissionInCurrentContext(&$updatedAccounts, $customer, NostoSDKAccount $account)
```
Comment (hannunosto) [inline classes/services/NostoCustomerService.php]: I don't think you need `$updatedAccounts` here. You can simply pass `$customer` & `$account` and return boolean. Then keep the `$updatedAccounts` up to date here: https://github.com/Nosto/nosto-prestashop/pull/248/files#diff-4e015f2b8e4768e084a8e34957fbd778R59. Also check the doc block & type hint for `$customer`

File: classes/services/NostoCustomerService.php [hunk 6]
```diff
@@ -39,38 +39,89 @@ class NostoCustomerService extends AbstractNostoService
      */
     public function customerUpdated($customer)
     {
-        if (!$customer->email) {
+        if (!$customer->email
```
Comment (hannunosto) [inline classes/services/NostoCustomerService.php]: Add type hint to the method signature -> `customerUpdated(Customer $customer)` & update doc block

File: classes/services/NostoCustomerService.php (hunk 3 above)
Comment (supercid) [reply classes/services/NostoCustomerService.php]: Fixed, thanks

File: classes/services/NostoCustomerService.php (hunk 4 above)
Comment (supercid) [reply classes/services/NostoCustomerService.php]: Fixed, thanks

File: classes/services/NostoCustomerService.php (hunk 5 above)
Comment (supercid) [reply classes/services/NostoCustomerService.php]: Good catch!

**Pull request merged**